google-oauth-client/src/main/java/com/google/api/client/auth/oauth2/JWTAuthentication.java

```diff
@@ -0,0 +1,127 @@
+ * TokenRequest#setClientAuthentication(HttpExecuteInterceptor)}.
+ *
+ * <p>
+ *     To use JWT authentication, grant_type must be "client_credentials".
+ *     If AuthorizationCodeTokenRequest.setGrantType() is called, set it to
+ *     JWTAuthentication.GRANT_TYPE_CLIENT_CREDENTIALS. It can also be left
+ *     uncalled. Setting it to any other value will cause an IllegalArgumentException.
+ * </p>
+ *
+ * <p>Sample usage:
+ *
+ * <pre>
+ * static void requestAccessToken() throws IOException {
+ * try {
+ * TokenResponse response = new AuthorizationCodeTokenRequest(new NetHttpTransport(),
+ * new JacksonFactory(), new GenericUrl("https://server.example.com/token")
+ * .setGrantType(JWTAuthentication.GRANT_TYPE_CLIENT_CREDENTIALS)
+ * .setClientAuthentication(
+ * new JWTAuthentication("eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzM4NCJ9...")).execute();
+ * System.out.println("Access token: " + response.getAccessToken());
+ * } catch (TokenResponseException e) {
+ * if (e.getDetails() != null) {
+ * System.err.println("Error: " + e.getDetails().getError());
+ * if (e.getDetails().getErrorDescription() != null) {
+ * System.err.println(e.getDetails().getErrorDescription());
+ * }
+ * if (e.getDetails().getErrorUri() != null) {
+ * System.err.println(e.getDetails().getErrorUri());
+ * }
+ * } else {
+ * System.err.println(e.getMessage());
+ * }
+ * }
+ * }
+ * </pre>
+ *
+ * <p>Implementation is immutable and thread-safe.
+ *
+ * @author Jun Ying
+ */
+
+public class JWTAuthentication
+    implements HttpRequestInitializer, HttpExecuteInterceptor {
+
+    public static final String GRANT_TYPE_KEY = "grant_type";
+
+    /** Predefined value for grant_type when using JWT **/
+    public static final String GRANT_TYPE_CLIENT_CREDENTIALS = "client_credentials";
+
+    public static final String CLIENT_ASSERTION_TYPE_KEY = "client_assertion_type";
+
+    /** Predefined value for client_assertion_type when using JWT **/
+    public static final String CLIENT_ASSERTION_TYPE = "urn:ietf:params:oauth:client-assertion-type:jwt-bearer";
+
+    public static final String CLIENT_ASSERTION_KEY = "client_assertion";
+
+    /** JWT for authentication. */
+    private final String jwt;
+
+    /**
+     * @param jwt JWT used for authentication
+     */
+    public JWTAuthentication(String jwt) {
+        this.jwt = Preconditions.checkNotNull(jwt);
+    }
+
+    public void initialize(HttpRequest request) throws IOException {
+        request.setInterceptor(this);
+    }
+
+    public void intercept(HttpRequest request) {
+        Map<String, Object> data = Data.mapOf(UrlEncodedContent.getContent(request).getData());
+        if (!data.containsKey(GRANT_TYPE_KEY)) {
+            data.put(GRANT_TYPE_KEY, GRANT_TYPE_CLIENT_CREDENTIALS);
+        } else {
+            String grantType = (String) data.get(GRANT_TYPE_KEY);
+            if (!grantType.equals(GRANT_TYPE_CLIENT_CREDENTIALS)) {
```

Comment:
cam grantType be null here?

Reply:
It's defensive in case the Request doesn't have setGrantType called. Since the value is predefined for JWT based authentication, it's safe to go ahead and set it if null.